### ISSUE TITLE
Parallel mini-batch scalability issue

### ISSUE THREAD

**Issue by rohskopf** [uber/horovod#8]
I am performing batch gradient descent, where the gradients are averaged across all training examples to move forward in a single training epoch. This can easily be parallelized by splitting the original batch into "mini-batches", and each process calculates gradients associated with its mini-batch. The resulting mini-batch gradients can be averaged on the master process to then move forward with training. This implementation should be "embarrassingly parallel" since the gradients of N examples are calculated over P processes in N/P mini-batches. A good animation is from TensorFlow's synchronous training examples:

![synchronous](https://user-images.githubusercontent.com/11083811/29498958-6bd6e0ee-85d4-11e7-83f6-b0e971ca380f.png)

I want to perform this type of training on a CPU cluster and I'm very familiar with MPI, so I think Horovod is a better choice than Distributed TensorFlow. I am having a hard time setting up this task, however. My recent attempt involves optimizing a single layer neural network to reproduce the parabola y = x^2 (just for simplicity):

```
import tensorflow as tf
import numpy as np
import horovod.tensorflow as hvd
import time

# Function 
def add_layer(inputs, in_size, out_size, activation_function=None):
    w = tf.Variable(tf.random_normal([in_size, out_size]))
    b = tf.Variable(tf.random_normal([1, out_size]))
    Wx_plus_b = tf.matmul(inputs, w) + b
    if activation_function is None:
        outputs = Wx_plus_b
    else:
        outputs = activation_function(Wx_plus_b)
    return outputs

# Initialize Horovod
hvd.init()
rank = hvd.rank()

# Make up some data
N = 20000
x_data = np.linspace(-1, 1, N)[:, np.newaxis]
y_data = np.square(x_data)

# Make mini-batches, each with N/procs elements
procs = hvd.size()
x_split = np.split(x_data, procs)
y_split = np.split(y_data, procs)
x_batch = x_split[rank]
y_batch = y_split[rank]

# Define placeholder for x_data and y_data
xs = tf.placeholder(tf.float32, [None, 1])
ys = tf.placeholder(tf.float32, [None, 1])
# Add hidden layer
l1 = add_layer(xs, 1, 3, activation_function=tf.tanh)
# Add output layer
prediction = add_layer(l1, 3, 1, activation_function=None)
# Define loss
loss = tf.reduce_mean(tf.reduce_sum(tf.square(ys-prediction), reduction_indices=[1]))
# Define optimizer
opt = tf.train.GradientDescentOptimizer(0.01)

# Add Horovod Distributed Optimizer
opt = hvd.DistributedOptimizer(opt)
# Compute the gradients for a list of variables.
grads_and_vars = opt.compute_gradients(loss)
# Ask the optimizer to apply the gradients
train_opt = opt.apply_gradients(grads_and_vars)

# Monitored session
hooks = [hvd.BroadcastGlobalVariablesHook(0)]
sess = tf.train.MonitoredTrainingSession(hooks=hooks)

# Initialize timer
if rank == 0:
    start = time.time()

# Perform the training
for i in range(10000):
    # Training step
    sess.run(train_opt, feed_dict={xs: x_batch, ys: y_batch})
    if i % 100 == 0:
        print("Rank: %d" % (rank), "Epoch: %d" % (i), "Loss: %.5f" % (sess.run(loss, feed_dict={xs: x_data, ys: y_data})) )

# Output the final time
if rank == 0:
    end = time.time()
    print("Elapsed time: %.5f sec" % (end - start) )
```

If this is saved in a file called `test.py`, you can run with:
`mpirun -np P python test.py`
where `P` is the number of processes to use. This parallelizes the gradient calculation from `N` training examples to `N/P` examples, and then the gradients are averaged using `hvd.allreduce` in the distributed optimizer.

There are 2 issues that I noticed:
(1) Even though the mini-batches are being parallelized properly, the training actually gets slower with increasing the number of processes. Is this simply due to overhead, since this is a rather simple example?
(2) There are `P` identical neural networks being trained simultaneously. I would like to have the separate processes calculate the gradients of their mini-batches, and send them back to the master process to proceed with training. Is there a way to set this up using Horovod?

I was thinking maybe (2) was causing (1), since many identical neural nets are being trained at once perhaps this is somehow slower?

Thanks for your time.

EDIT: Cleaned up code for more clarity.

**Comment by alsrgv** [comment]
Hi @rohskopf,

Your code is correct, but you need to make your network more complicated to get wall-clock speedup from distributed training.  The rule of thumb we found is that one step should take about one second.  You can try to get there by increasing batch size per worker until you reach such speed.  I had to increase N to `40000000` to see actual scaling, which is too large batch size for this problem.

In practice you'd also want to randomize how your input data is fed into network.  One common use case is to randomly read slices of input data in random order via combination of [filename shuffling](https://www.tensorflow.org/programmers_guide/reading_data#filenames_shuffling_and_epoch_limits) and [batch shuffling](https://www.tensorflow.org/programmers_guide/reading_data#batching).

I'm not sure I completely understand second issue.  Your script is already accomplishing this, with the caveat that instead of sending gradients to master process, they are distributed to all processes.

Thanks,
Alex

**Comment by rohskopf** [comment]
Thanks for the reply @alsrgv .

I noticed the same thing with the scaling. It makes sense that it'll scale well with more complicated networks.

Regarding the second issue - I noticed that identical losses are printed for each process during the training. This is because the averaged gradients are being sent to all processes, like you mentioned. And then all processes apply the gradients in the training step. I was wondering if it could speed things up if _only_ the master process does the training (applying the gradients) while the other processes simply calculate the gradients and send them to the master for averaging and applying.

**Comment by alsrgv** [comment]
With single master responsible for averaging it's highly likely that it will become a bottleneck.

This brings interesting point.  TensorFlow uses multi-threading on CPU, so if you have multi-core CPU you'd already compute gradients and average them in separate threads.

However, when you have GPU, we found that sometimes it's good to let the GPU work on computing gradients and let the CPU do the averaging.  One example of such network is Inception V3 - it has large amount of computation required, and not too many parameters to exchange (and average).
 [README](https://github.com/uber/horovod#have-gpus) touches on this subject and describes how DistributedOptimizer can be advised to use CPU for averaging.

**Comment by rohskopf** [comment]
Very interesting. Thanks again for your help.

**Comment by alsrgv** [comment]
I will close this issue, please feel free to reopen if you have additional questions.

**Comment by nischalhp** [comment]
@alsrgv We are facing the same issue with our implementation of a deep learning architecture with keras and running horovod on it. Could you point me to the example where you are using CPU to computing the average by the DistirbutedOptimizer?

**Comment by alsrgv** [comment]
@nischalhp, do you want to compute all the averages on CPU, or just for specific gradients?  In the former case, you can specify `hvd.DistributedOptimizer(opt, device_dense='/cpu:0')`.

**Comment by nischalhp** [comment]
@alsrgv Okay, thank you. Will try this out and keep you posted. I think we can close this issue for now.

**Comment by alsrgv** [comment]
@nischalhp, sounds good.
